rfcs/00x-frontend-library-split.md

```diff
@@ -0,0 +1,88 @@
+- Feature Name: frontend-library-split
+- Start Date: 2020-11-23
+- RFC PR: [amundsen-io/rfcs#150(https://github.com/amundsen-io/rfcs/pull/15) (after opening the RFC PR, update this with a link to it and update the file name)
+- Amundsen Issue: [amundsen-io/amundsen#0000](https://github.com/amundsen-io/amundsen/issues/0000) (leave this empty for now)
+
+# Split of the frontendlibrary repository in two
+
+## Summary
+
+At this moment, our frontend repository contains a Flask application that acts as an API server and also serves a frontend application based in React.js. This means that there are two applications (Flask and Frontend), with two languages (Python and JavaScript), two configuration settings, two dependency management systems (pip and npm), two sets of owners (frontend and backend developers), and two main "responsibilities" (API server and User Interface application) baked into the same repository. This is a problem, as it makes the system complex, couples both systems and stifles the evolution of both applications.
+
+We propose to split our current amundsenfrontend repository into a backend-only API server and a frontend-only application.
+
+## Motivation
+
+> Why are we doing this? What use cases does it support? What is the expected outcome?
+
+The main benefits of this proposal are:
+
+- Easier evolution of the applications
+- Allow for multiple frontends
+- Streamlined development
+
+### Evolution of applications
+
+Having two applications in different repositories will reduce the coupling, allowing us to evolve both apps independently, implementing the newest best practices on both. For example, this would make it easier to create a GraphQL based interface for our API, as well as adopting a full stack structure based in Next.js or a static site based in Gatsby in the frontend.
+
+### Allow for multiple frontends
+
+Extracting the API server into its own applications would allow us to create different frontends for the Amundsen application. Think about creating a mobile application, a desktop app or forking the current UI application to customize it deeply to the user's specifications.
```

Comment:
It would be indeed useful to know how many users have forked completely their own frontend library without working on top of the standard. Our current configuration options help a bit with the customization, but they can only go so far.

Reply:
There's only one company in the community who has created their own - to match their own look and feel. That's Cimpress (video and notes [here](https://docs.google.com/document/d/1MT3qd_YjFiA17en94IEZcWaEgjdcaLhhZb-mX1UyWF8/edit#)). I don't think it's a super common use-case, but I can double check if you'd like.

There are others that have added some minor to moderate changes to UI that haven't been merged to master. On minor changes, examples are REA group and iRobot adding new structured metadata columns (for things like primary key, or link to example Mode reports respectively), and Square adding a human in the loop to approve/reject suggested PII tags ([PDF doc in slack](https://amundsenworkspace.slack.com/files/UJ35GMPUZ/FSR2GE141/square_demo___amundsen.pdf), [blog](https://developer.squareup.com/blog/using-amundsen-to-support-user-privacy-via-metadata-collection-at-square/)). They have forked the UI because they think our focus is on data discovery, while their use-cases are more in privacy engineering/data governance. I think that's reasonable.

So, all in all, I think it's super rare. My potentially controversial opinion: I won't over-index on the forking use-case. Thoughts?

Reply:
Thanks for the explanation @markgrover!